### PR TITLE
MMAP index for single-chunk indexes

### PR DESCRIPTION
**Pull request by udingankar**

### CODE REVIEW

File: colbert/indexing/codecs/residual_embeddings.py [hunk 1]
```diff
@@ -29,28 +28,50 @@ def load_chunks(cls, index_path, chunk_idxs, num_embeddings):
 
         dim, nbits = get_dim_and_nbits(index_path)
 
-        codes = torch.empty(num_embeddings, dtype=torch.int32)
-        residuals = torch.empty(num_embeddings, dim // 8 * nbits, dtype=torch.uint8)
+        print_message("#> Loading codes and residuals...")
 
-        codes_offset = 0
+        if len(chunk_idxs) == 1: # Memory mapping is enabled for single-file indexes
```
Comment (santhnm2) [inline colbert/indexing/codecs/residual_embeddings.py:33]: Can you add comments explaining what the constants are for? In particular, the 80 and 320

File: colbert/indexing/codecs/residual_embeddings.py [hunk 2]
```diff
@@ -29,28 +28,50 @@ def load_chunks(cls, index_path, chunk_idxs, num_embeddings):
...
+                codes[codes_offset:codes_endpos] = chunk.codes
+                residuals[codes_offset:codes_endpos] = chunk.residuals
+
+                codes_offset = codes_endpos
 
         return cls(codes, residuals)
 
+    @classmethod
+    def load_from_single_index(self):
+        path = "/future/u/udingank/ColBERT/experiments/default/indexes/msmarco_single.nbits=2.latest/"
```
Comment (santhnm2) [inline colbert/indexing/codecs/residual_embeddings.py:70]: Replace the hard-coded path

**Review by santhnm2 — CHANGES_REQUESTED** [review]
Looks good % a couple small changes, we should maybe add some message after indexing and/or while searching to coalesce their index if they want to enable the MMAP path

**Comment by santhnm2** [comment]
Can we also add a test for this?

File: colbert/indexing/codecs/residual_embeddings.py (hunk 2 above)
Comment (udingankar) [reply colbert/indexing/codecs/residual_embeddings.py:70]: load_from_single_index() is actually a stale function, going to get rid of it completely.

**Comment by okhat** [comment]
Yeah we should check that LoTTE Lifestyle:dev and Pooled Success@5 or MS MARCO MRR are still at the same score level that they used to be. Btw is this optional or does it replace the old index loading approach?

**Comment by okhat** [comment]
Also, thanks a ton @udingankar for the exciting new developments!

File: colbert/indexing/codecs/residual_embeddings.py (hunk 1 above)
Comment (udingankar) [reply colbert/indexing/codecs/residual_embeddings.py:33]: Done

**Comment by udingankar** [comment]
@okhat - I've tested mmap on MS MARCO v1 and double checked that the MRR@10 matches the control.
It's enabled for indexes that contain only 1 chunk by default - I could make mmap optional however.

**Comment by okhat** [comment]
Okay this sounds good

**Comment by udingankar** [comment]
@santhnm2 - added a unit test for the coalesced index to https://github.com/stanford-futuredata/ColBERT/pull/205

**Comment by okhat** [comment]
Can we make this "optional" (and False/OFF by default) and then figure out merging it?

All ok from my end as long as the OFF case is identical to the default workflow. This seems like a great feature to have.

**Comment by santhnm2** [comment]
I think we just need @VThejas to sign off that everything worked as expected for his testing - I think we had also discussed some minor tweaks so I'll follow up on the status of those.

**Pull request merged**